File: fasta_lake/dianovo.py
"""Read complete DIANovo predictions without guessing across unresolved mass gaps.

DIANovo: https://github.com/hearthewind/dianovo (cc413f25140e47a5f6a4234bae15da0ef525ac3a),
``denovo_from_feature_detection/rnova/task/sequence_generation_inference_utils.py``
and ``optimal_path_inference_classic.py``. These are FastaLake input checks, not
DIANovo inference. Sequence generation writes one probability per residue; the
space-delimited path output prepends a zero marker. See docs/how-to/de-novo-inputs.md.
The matching Rust implementation is rust/common/dianovo.rs.
"""

import math

_AMINO_ACIDS = frozenset("ACDEFGHIKLMNPQRSTVWY")


def parse_dianovo_prediction(sequence: str, probabilities: str) -> tuple[str, float]:
    """Return a complete uppercase peptide and mean confidence, including zeros.

    Accept lowercase residues and whitespace between residues. Reject mass gaps,
    noncanonical residues, malformed lists, nonfinite/out-of-range probabilities
    and length mismatches with ValueError. Only a space-delimited vector with
    exactly one extra leading zero is interpreted as a path-marker vector.
    All-zero residue vectors are valid here; positive-score selection excludes them.
    """
    raw_sequence = "".join(c for c in sequence if c not in " \t\r\n\v\f")
    if not raw_sequence.isascii():
        raise ValueError("unresolved_gap_or_noncanonical_sequence")
    peptide = raw_sequence.upper()
    if not peptide or any(aa not in _AMINO_ACIDS for aa in peptide):
        raise ValueError("unresolved_gap_or_noncanonical_sequence")
    raw = probabilities.strip()
    bracketed = raw.startswith("[") and raw.endswith("]")
    if bracketed:
        raw = raw[1:-1].strip()
    comma_format = bracketed or "," in raw
    tokens = raw.split(",") if comma_format else raw.split()
    if any(not token.isascii() or "_" in token for token in tokens):
        raise ValueError("invalid_probability_vector")
    try:
        values = [float(token.strip()) for token in tokens]
    except ValueError as exc:
        raise ValueError("invalid_probability_vector") from exc
    if not values or any(not math.isfinite(p) or not 0 <= p <= 1 for p in values):
        raise ValueError("invalid_probability_vector")
    if not comma_format and len(values) == len(peptide) + 1 and values[0] == 0:
        values = values[1:]
    if len(values) != len(peptide):
        raise ValueError("sequence_probability_length_mismatch")
    # Explicit left-to-right f64 addition matches Rust's ranking at cutoff ties.
    total = 0.0
    for value in values:
        total += value
    return peptide, total / len(values)

File: rust/common/dianovo.rs
//! FastaLake input checks for DIANovo; this is not model inference.
//! Upstream: https://github.com/hearthewind/dianovo at cc413f25140e47a5f6a4234bae15da0ef525ac3a.
//! See denovo_from_feature_detection/rnova/task/{sequence_generation_inference_utils,
//! optimal_path_inference_classic}.py and docs/how-to/de-novo-inputs.md.
//! Keep semantics aligned with fasta_lake/dianovo.py; cross-language tests compare
//! extraction and razor selection with independently scored generic predictions.

/// Parse a complete canonical peptide and its mean confidence, retaining zeros.
/// The space-delimited path vector alone may have one extra leading zero marker.
/// Never erase numeric mass gaps or silently discard malformed probabilities.
pub fn parse_prediction(sequence: &str, probabilities: &str) -> Option<(String, f64)> {
    let peptide: String = sequence
        .chars()
        .filter(|c| !c.is_ascii_whitespace())
        .map(|c| c.to_ascii_uppercase())
        .collect();
    if peptide.is_empty()
        || !peptide
            .bytes()
            .all(|aa| b"ACDEFGHIKLMNPQRSTVWY".contains(&aa))
    {
        return None;
    }
    let raw = probabilities.trim();
    let bracketed = raw.starts_with('[') && raw.ends_with(']');
    let raw = if bracketed {
        &raw[1..raw.len() - 1]
    } else {
        raw
    };
    let comma_format = bracketed || raw.contains(',');
    let tokens: Vec<&str> = if comma_format {
        raw.split(',').collect()
    } else {
        raw.split_whitespace().collect()
    };
    let values: Vec<f64> = tokens
        .iter()
        .map(|token| token.trim().parse::<f64>().ok())
        .collect::<Option<_>>()?;
    if values.is_empty()
        || values
            .iter()
            .any(|p| !p.is_finite() || !(0.0..=1.0).contains(p))
    {
        return None;
    }
    let residues = if !comma_format && values.len() == peptide.len() + 1 && values[0] == 0.0 {
        &values[1..]
    } else {
        &values[..]
    };
    if residues.len() != peptide.len() {
        return None;
    }
    let score = residues.iter().sum::<f64>() / residues.len() as f64;
    Some((peptide, score))
}

#[cfg(test)]
mod tests {
    use super::parse_prediction;

    #[test]
    fn zeros_and_path_markers_have_different_meanings() {
        assert_eq!(
            parse_prediction("AcD", "[0, 0.9, 0.9]"),
            Some(("ACD".into(), 0.6))
        );
        assert_eq!(
            parse_prediction("A c D", "0 0 0.9 0.9"),
            Some(("ACD".into(), 0.6))
        );
        assert_eq!(
            parse_prediction("AcD", "0 0.9 0.9"),
            Some(("ACD".into(), 0.6))
        );
        assert_eq!(
            parse_prediction("AcD", "[0, 0, 0]"),
            Some(("ACD".into(), 0.0))
        );
    }

    #[test]
    fn no_gap_joining_or_partial_vector_recovery() {
        for (seq, prob) in [
            ("AC113.084D", "[0.9, 0.9, 0.9]"),
            ("AXB", "[1, 1, 1]"),
            ("ACD", "[0, 1, 1, 1]"),
            ("ACD", "0.1 1 1 1"),
            ("ACD", "[1, bogus, 1]"),
            ("ACD", "[1, NaN, 1]"),
            ("ACD", "[1, inf, 1]"),
            ("ACD", "[1, -0.1, 1]"),
            ("ACD", "[1, 1.1, 1]"),
            ("ACD", "[1, 1, 1"),
            ("ACD", "[1, 1, 1,]"),
            ("ACD", "[]"),
        ] {
            assert_eq!(parse_prediction(seq, prob), None, "{seq}: {prob}");
        }
    }
}
